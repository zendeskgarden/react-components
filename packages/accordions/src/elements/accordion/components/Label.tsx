/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { forwardRef, HTMLAttributes } from 'react';
import { StyledButton } from '../../../styled';
import { useAccordionContext, useHeaderContext, useSectionContext } from '../../../utils';

export const Label = forwardRef<HTMLButtonElement, HTMLAttributes<HTMLButtonElement>>(
  (props, ref) => {
    const sectionIndex = useSectionContext();
    const { isCompact, isCollapsible, expandedSections } = useAccordionContext();
    const isExpanded = expandedSections.includes(sectionIndex);
    const { isHovered, otherTriggerProps } = useHeaderContext();

    return (
      <StyledButton
        ref={ref}
        isCompact={isCompact}
        isHovered={isHovered}
        isExpanded={isExpanded}
        isCollapsible={isCollapsible}
        {...otherTriggerProps}
        {...props}
      />
    );
  }
);

Label.displayName = 'Label';
