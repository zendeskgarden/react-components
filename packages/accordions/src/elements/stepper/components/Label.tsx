/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import CheckCircleStrokeIcon from '@zendeskgarden/svg-icons/src/16/check-sm-stroke.svg';
import { IStepperLabelProps } from '../../../types';
import { StyledLabel, StyledLabelText, StyledIcon, StyledIconFlexContainer } from '../../../styled';
import { useStepContext } from '../../../utils';

const LabelComponent = forwardRef<HTMLDivElement, IStepperLabelProps>(
  ({ icon, iconProps, isHidden, children, ...other }, ref) => {
    const { currentStepIndex, isActive, isCompleted, isHorizontal } = useStepContext();
    const numericStep = currentStepIndex + 1;
    const stepIcon = icon || numericStep;

    const styledIcon = (
      <StyledIcon $isActive={isActive} $isHorizontal={isHorizontal}>
        {isCompleted ? <CheckCircleStrokeIcon {...iconProps} /> : stepIcon}
      </StyledIcon>
    );

    return (
      <StyledLabel ref={ref} $isActive={isActive} $isHorizontal={isHorizontal} {...other}>
        {isHorizontal ? (
          <StyledIconFlexContainer>{styledIcon}</StyledIconFlexContainer>
        ) : (
          styledIcon
        )}
        <StyledLabelText $isHidden={isHidden} $isHorizontal={isHorizontal}>
          {children}
        </StyledLabelText>
      </StyledLabel>
    );
  }
);

LabelComponent.displayName = 'Stepper.Label';

LabelComponent.propTypes = {
  icon: PropTypes.any,
  iconProps: PropTypes.object,
  isHidden: PropTypes.bool
};

/**
 * @extends HTMLAttributes<HTMLDivElement>
 */
export const Label = LabelComponent;
