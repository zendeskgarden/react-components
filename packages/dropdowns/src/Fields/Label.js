/* eslint-disable react/prop-types */
/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { composeEventHandlers } from '@zendeskgarden/container-selection';
import useDropdownContext from '../utils/useDropdownContext';
import useFieldContext from '../utils/useFieldContext';
import { StyledLabel } from '../styled';

/**
 * Accepts all `<label>` props. Must be nested with a `<Field>` component.
 */
const Label = React.forwardRef(({ onMouseEnter, onMouseLeave, ...other }, ref) => {
  const {
    downshift: { getLabelProps }
  } = useDropdownContext();
  const { setIsLabelHovered } = useFieldContext();

  const labelProps = getLabelProps({
    onMouseEnter: composeEventHandlers(onMouseEnter, () => {
      setIsLabelHovered(true);
    }),
    onMouseLeave: composeEventHandlers(onMouseLeave, () => {
      setIsLabelHovered(false);
    }),
    ...other
  });

  return <StyledLabel ref={ref} {...labelProps} />;
});

Label.propTypes = {
  regular: PropTypes.bool,
  small: PropTypes.bool
};

export default Label;
