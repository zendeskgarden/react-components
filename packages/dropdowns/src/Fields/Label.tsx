/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLProps } from 'react';
import PropTypes from 'prop-types';
import { composeEventHandlers } from '@zendeskgarden/container-utilities';
import useDropdownContext from '../utils/useDropdownContext';
import useFieldContext from '../utils/useFieldContext';
import { StyledLabel } from '../styled';

interface ILabelProps extends HTMLProps<HTMLLabelElement> {
  regular?: boolean;
  small?: boolean;
}

/**
 * Accepts all `<label>` props. Must be nested with a `<Field>` component.
 */
const Label = React.forwardRef<HTMLLabelElement, ILabelProps>(
  ({ onMouseEnter, onMouseLeave, ...other }, ref) => {
    const {
      downshift: { getLabelProps }
    } = useDropdownContext();
    const { setIsLabelHovered } = useFieldContext();

    const labelProps = getLabelProps({
      onMouseEnter: composeEventHandlers(onMouseEnter, () => {
        setIsLabelHovered(true);
      }),
      onMouseLeave: composeEventHandlers(onMouseLeave, () => {
        setIsLabelHovered(false);
      }),
      ...other
    });

    return <StyledLabel ref={ref} {...labelProps} />;
  }
) as React.FunctionComponent<ILabelProps>;

Label.propTypes = {
  regular: PropTypes.bool,
  small: PropTypes.bool
};

export default Label;
