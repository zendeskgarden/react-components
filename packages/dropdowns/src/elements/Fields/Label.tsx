/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { composeEventHandlers } from '@zendeskgarden/container-utilities';
import { Label as FormLabel } from '@zendeskgarden/react-forms';
import useDropdownContext from '../../utils/useDropdownContext';
import useFieldContext from '../../utils/useFieldContext';

interface ILabelProps extends HTMLAttributes<HTMLLabelElement> {
  isRegular?: boolean;
}

/**
 * Accepts all `<label>` props. Must be nested with a `<Field>` component.
 */
export const Label = React.forwardRef<HTMLLabelElement, ILabelProps>(
  ({ onMouseEnter, onMouseLeave, ...other }, ref) => {
    const {
      downshift: { getLabelProps }
    } = useDropdownContext();
    const { setIsLabelHovered } = useFieldContext();

    const labelProps = getLabelProps({
      onMouseEnter: composeEventHandlers(onMouseEnter, () => {
        setIsLabelHovered(true);
      }),
      onMouseLeave: composeEventHandlers(onMouseLeave, () => {
        setIsLabelHovered(false);
      }),
      ...other
    });

    return <FormLabel ref={ref} {...labelProps} />;
  }
);

Label.displayName = 'Label';

Label.propTypes = {
  isRegular: PropTypes.bool
};
