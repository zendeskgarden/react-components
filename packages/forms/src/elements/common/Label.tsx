/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { composeEventHandlers } from '@zendeskgarden/container-utilities';
import useFieldContext from '../../utils/useFieldContext';
import useInputContext from '../../utils/useInputContext';
import {
  StyledLabel,
  StyledCheckLabel,
  StyledCheckSvg,
  StyledDashSvg,
  StyledRadioLabel,
  StyledRadioSvg,
  StyledToggleLabel,
  StyledToggleSvg
} from '../../styled';

export interface ILabelProps extends HTMLAttributes<HTMLLabelElement> {
  /** Applies regular (non-bold) font weight */
  isRegular?: boolean;
}

/**
 * Must be rendered within a `<Field>` element; accepts all `<label>` attributes
 * and events.
 */
export const Label = React.forwardRef<HTMLLabelElement, ILabelProps>((props, ref) => {
  const fieldContext = useFieldContext();
  const type = useInputContext();

  let combinedProps = props;

  if (fieldContext) {
    combinedProps = fieldContext.getLabelProps(combinedProps);

    if (type === undefined) {
      const { setIsLabelActive, setIsLabelHovered, multiThumbRangeRef } = fieldContext;

      combinedProps = {
        ...combinedProps,
        onMouseUp: composeEventHandlers(props.onMouseUp, () => {
          setIsLabelActive(false);
        }),
        onMouseDown: composeEventHandlers(props.onMouseDown, () => {
          setIsLabelActive(true);
        }),
        onMouseEnter: composeEventHandlers(props.onMouseEnter, () => {
          setIsLabelHovered(true);
        }),
        onMouseLeave: composeEventHandlers(props.onMouseLeave, () => {
          setIsLabelHovered(false);
        }),
        onClick: composeEventHandlers(props.onClick, () => {
          multiThumbRangeRef.current && multiThumbRangeRef.current.focus();
        })
      };
    }
  }

  if (type === 'radio') {
    return (
      <StyledRadioLabel ref={ref} {...(combinedProps as any)}>
        <StyledRadioSvg />
        {props.children}
      </StyledRadioLabel>
    );
  } else if (type === 'checkbox') {
    /**
     * `onLabelSelect` is a workaround for checkbox label `shift + click` bug in Firefox
     * See: https://bugzilla.mozilla.org/show_bug.cgi?id=559506
     */
    const onLabelSelect = (e: React.KeyboardEvent<HTMLInputElement>) => {
      const isFirefox = navigator.userAgent.toLowerCase().indexOf('firefox') > -1;

      if (fieldContext && isFirefox && e.target instanceof Element) {
        const inputId = e.target.getAttribute('for');

        if (!inputId) return;

        const input = document.getElementById(inputId) as HTMLInputElement | null;

        if (input && input.type === 'checkbox') {
          if (e.shiftKey) {
            input.click();
            input.checked = true;
          }
          input.focus();
        }
      }
    };

    combinedProps = {
      ...combinedProps,
      onClick: composeEventHandlers(combinedProps.onClick, onLabelSelect)
    };

    return (
      <StyledCheckLabel ref={ref} {...(combinedProps as any)}>
        <StyledCheckSvg />
        <StyledDashSvg />
        {props.children}
      </StyledCheckLabel>
    );
  } else if (type === 'toggle') {
    return (
      <StyledToggleLabel ref={ref} {...(combinedProps as any)}>
        <StyledToggleSvg />
        {props.children}
      </StyledToggleLabel>
    );
  }

  return <StyledLabel ref={ref} {...(combinedProps as any)} />;
});

Label.displayName = 'Label';

Label.propTypes = {
  isRegular: PropTypes.bool
};
