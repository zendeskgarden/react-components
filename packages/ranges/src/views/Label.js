/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import PropTypes from 'prop-types';
import styled from 'styled-components';
import classNames from 'classnames';
import { retrieveTheme, isRtl } from '@zendeskgarden/react-theming';
import RangeStyles from '@zendeskgarden/css-forms/dist/range.css';

const COMPONENT_ID = 'ranges.label';

/**
 * Accepts all `<label>` props
 */
const Label = styled.label.attrs({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  className: props =>
    classNames(RangeStyles['c-range__label'], {
      [RangeStyles['c-range__label--regular']]: props.regular,
      [RangeStyles['c-range__label--sm']]: props.small,

      // RTL
      [RangeStyles['is-rtl']]: isRtl(props)
    })
})`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

Label.propTypes = {
  regular: PropTypes.bool,
  small: PropTypes.bool
};

/** @component */
export default Label;
