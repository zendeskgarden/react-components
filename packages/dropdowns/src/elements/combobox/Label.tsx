/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import { composeEventHandlers } from '@zendeskgarden/container-utilities';
import useFieldContext from '../../context/useFieldContext';
import { ILabelProps } from '../../types';
import { StyledLabel } from '../../views';

/**
 * @deprecated use `Field.Label` instead
 *
 * @extends LabelHTMLAttributes<HTMLLabelElement>
 */
export const Label = forwardRef<HTMLLabelElement, ILabelProps>(
  ({ onClick, onMouseEnter, onMouseLeave, ...props }, ref) => {
    const { labelProps } = useFieldContext();

    return (
      <StyledLabel
        {...labelProps}
        onClick={composeEventHandlers(onClick, labelProps?.onClick)}
        onMouseEnter={composeEventHandlers(onMouseEnter, labelProps?.onMouseEnter)}
        onMouseLeave={composeEventHandlers(onMouseLeave, labelProps?.onMouseLeave)}
        {...props}
        ref={ref}
      />
    );
  }
);

Label.displayName = 'Label';

Label.propTypes = {
  hidden: PropTypes.bool,
  isRegular: PropTypes.bool
};
