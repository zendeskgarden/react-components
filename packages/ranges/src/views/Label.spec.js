/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { render, renderRtl } from 'garden-test-utils';

import Label from './Label';

describe('Label', () => {
  it('renders default styling', () => {
    const { container } = render(<Label />);

    expect(container.firstChild).toHaveClass('c-range__label');
  });

  it('renders RTL styling', () => {
    const { container } = renderRtl(<Label />);

    expect(container.firstChild).toHaveClass('is-rtl');
  });

  it('renders small styling if provided', () => {
    const { container } = render(<Label small />);

    expect(container.firstChild).toHaveClass('c-range__label--sm');
  });

  it('renders regular styling if provided', () => {
    const { container } = render(<Label regular />);

    expect(container.firstChild).toHaveClass('c-range__label--regular');
  });
});
