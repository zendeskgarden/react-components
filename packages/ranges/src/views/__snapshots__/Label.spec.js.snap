// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label renders RTL styling 1`] = `
<label
  className="Label-bxMjsr c-range__label is-rtl jVZUdW"
  data-garden-id="ranges.label"
  data-garden-version="version"
/>
`;

exports[`Label renders default styling 1`] = `
<label
  className="Label-bxMjsr c-range__label jVZUdW"
  data-garden-id="ranges.label"
  data-garden-version="version"
/>
`;

exports[`Label renders regular styling if provided 1`] = `
<label
  className="Label-bxMjsr c-range__label c-range__label--regular jVZUdW"
  data-garden-id="ranges.label"
  data-garden-version="version"
/>
`;

exports[`Label renders small styling if provided 1`] = `
<label
  className="Label-bxMjsr c-range__label c-range__label--sm jVZUdW"
  data-garden-id="ranges.label"
  data-garden-version="version"
/>
`;
